src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,337 @@
+ *
+ * This software is Copyright (c) 2015, Daniel Mende <dmende [at] ernw.de> and
+ * Copyright (c) 2017, Dhiru Kholia <dhiru [at] openwall.com>, and it is hereby
+ * released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ *  * Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following disclaimer
+ *    in the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ *  * Neither the name of the  nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ *    A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ *    OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ *    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ *    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ *    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_tacacsplus;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_tacacsplus);
+#else
+
+#include <string.h>
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               2048
+#endif
+#endif
+
+#include "formats.h"
+#include "misc.h"
+#include "common.h"
+#include "params.h"
+#include "options.h"
+#include "md5.h"
+#include "johnswap.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "tacacs-plus"
+#define FORMAT_NAME             "TACACS+"
+#define FORMAT_TAG              "$tacacs-plus$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "MD5 32/" ARCH_BITS_STR
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        0
+#define PLAINTEXT_LENGTH        125
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            1
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define SALT_ALIGN              sizeof(uint32_t)
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
```

Comment:
Having `MAX_KEYS_PER_CRYPT` at 1 is almost certainly wrong, since this looks like a fast format and since `OMP_SCALE` is so high. Instead, of 1 and 2048, we should use something like 64 and 32, respectively - need to tune the split, but generally the smallest sensible `MAX_KEYS_PER_CRYPT` for fast formats is 64. See: http://www.openwall.com/lists/john-dev/2017/11/12/1

It'd be nice to get this right at least for new formats. Would be less work for us to do when we approach the tuning of these things.